src/hotspot/share/runtime/vmThread.cpp

```diff
@@ -381,118 +291,188 @@ class HandshakeALotClosure : public HandshakeClosure {
+  } else if (SafepointALot) {
+    _next_vm_operation = &safepointALot_op;
   }
-  // Nothing to be done.
-  return NULL;
 }
 
-void VMThread::loop() {
-  assert(_cur_vm_operation == NULL, "no current one should be executing");
+bool VMThread::set_next_operation(VM_Operation *op) {
+  if (_next_vm_operation != NULL) {
+    return false;
+  }
+  log_debug(vmthread)("Adding VM operation: %s", op->name());
 
-  SafepointSynchronize::init(_vm_thread);
+  _next_vm_operation = op;
 
-  while(true) {
-    //
-    // Wait for VM operation
-    //
-    // use no_safepoint_check to get lock without attempting to "sneak"
-    { MonitorLocker mu_queue(VMOperationQueue_lock,
-                             Mutex::_no_safepoint_check_flag);
-
-      // Look for new operation
-      assert(_cur_vm_operation == NULL, "no current one should be executing");
-      _cur_vm_operation = _vm_queue->remove_next();
-
-      while (!should_terminate() && _cur_vm_operation == NULL) {
-        // wait with a timeout to guarantee safepoints at regular intervals
-        // (if there is cleanup work to do)
-        (void)mu_queue.wait(GuaranteedSafepointInterval);
-
-        // Support for self destruction
-        if ((SelfDestructTimer != 0) && !VMError::is_error_reported() &&
-            (os::elapsedTime() > (double)SelfDestructTimer * 60.0)) {
-          tty->print_cr("VM self-destructed");
-          exit(-1);
-        }
-
-        // If the queue contains a safepoint VM op,
-        // clean up will be done so we can skip this part.
-        if (!_vm_queue->peek_at_safepoint_priority()) {
-
-          // Have to unlock VMOperationQueue_lock just in case no_op_safepoint()
-          // has to do a handshake when HandshakeALot is enabled.
-          MutexUnlocker mul(VMOperationQueue_lock, Mutex::_no_safepoint_check_flag);
-          if ((_cur_vm_operation = VMThread::no_op_safepoint()) != NULL) {
-            // Force a safepoint since we have not had one for at least
-            // 'GuaranteedSafepointInterval' milliseconds and we need to clean
-            // something. This will run all the clean-up processing that needs
-            // to be done at a safepoint.
-            SafepointSynchronize::begin();
-            SafepointSynchronize::end();
-            _cur_vm_operation = NULL;
-          }
-        }
-        _cur_vm_operation = _vm_queue->remove_next();
-      }
+  HOTSPOT_VMOPS_REQUEST(
+                   (char *) op->name(), strlen(op->name()),
+                   op->evaluate_at_safepoint() ? 0 : 1);
+  return true;
+}
+
+void VMThread::until_executed(VM_Operation* op) {
+  MonitorLocker ml(VMOperation_lock,
+                   Thread::current()->is_Java_thread() ?
+                     Mutex::_safepoint_check_flag :
+                     Mutex::_no_safepoint_check_flag);
+  while (true) {
+    if (VMThread::vm_thread()->set_next_operation(op)) {
+      ml.notify_all();
+      break;
+    }
+    ml.notify_all();
+    ml.wait();
```

Comment:
It is really odd to see back-to-back `notify_all` and `wait`. Would that mean two threads sitting in this loop would just wake one another continuously?

Reply:
This can theoretically be an issue if there is operation set and then two non-JavaThread trying set a operation.
Or if there is a handshake operation set.

I did not consider this scenario proper.
I think we need two separate Monitors.
Fixing

Reply:
Please go over the changes in fceebea and see if you agree that it is correct :)
Passes t1-t8.